File: C/projects/CARDAMOM_MODELS/DALEC/DALEC_ALL/KNORR_ALLOCATION.c
#pragma once
#include "./LAI_KNORR_funcs.c"

typedef struct {
    struct{
        double temp; // mean air temperature (K)
        double deltat;
        double latitude;
        double DOY;
        double lambda;
        double lambda_max;
        double T_phi;
        double T_r;
        double plgr;
        double k_L;
        double pasm; 
        double ET;
        double tau_W;
        double t_c;
        double t_r;
        double T_memory; //udpated each iteration (automatically), can also be externally prescribed, no functional difference.
        double lambda_max_memory; //updated each iteration (automatically), can also be externally prescribed, no functional difference.
    }IN;
    struct {
        double lambda_next;
        double T;
        double laim;
        double dlambdadt;
        double  f_T;
        double f_d;
    }OUT;
    
    
    
}KNORR_ALLOCATION_STRUCT;


/* Leaf Phenology Model of Knorr et al. (2010, doi: 10.1029/2009JG001119)*/
//double* KNORR_ALLOCATION(double const *met_list, double const *var_list)
int KNORR_ALLOCATION(KNORR_ALLOCATION_STRUCT * K)
{

  /*initialize intermediate variables*/
  double daylengthpars[3];
  double daylength;
  double n, deltat;
  double meantemp;
  double t_c, t_r, f_d, td_deviation;
  double T_init, T_phi, T_r;
  double T, T_memory, T_deviation, f_T, tau_m;
  double plgr, k_L;
  double f, r;
  double pasm, ET;
  double lambda, lambda_W, lambda_tilde_max, lambda_max;
  double dlambdadt, lambda_lim, laim, lambda_max_memory;
  double tau_W, tau_s;
  double lambda_next;

  
  
  n=1.0; /*(double)var_list[0];*/
  deltat=(double)K->IN.deltat;
  lambda=(double)K->IN.lambda;
  T_init=(double)0.0;
  T_phi=K->IN.T_phi ;  // mean temp at leaf onset (K)
  T_r=K->IN.T_r;       // spatial range of T_phi (K)
  T_memory=K->IN.T_memory; // temperature memory (K)
  meantemp=K->IN.temp;     // mean surface air temperature (K)
   tau_m=1.0; //var_list[6];
   plgr=K->IN.plgr;//
   k_L=K->IN.k_L;
   pasm=K->IN.pasm;
   ET=K->IN.ET;
   lambda_max=K->IN.lambda_max;
   tau_W=K->IN.tau_W;
   lambda_max_memory=K->IN.lambda_max_memory;  //TIME-DEPENDENT I.E. SAVE IN MEMORY
   tau_s=30.0; // Averaging period for water/structural memory (units of days; must be in same units as deltat), usually kept constant but could potentially be a tuneable parameter
   t_c=K->IN.t_c; 
   t_r=K->IN.t_r; 

  
  daylengthpars[0]=K->IN.latitude;  /* latitude */
  daylengthpars[1]=K->IN.DOY;  /* day of year */
  daylengthpars[2]=DGCM_PI;  /* pi */

  /* Initialization: only run this on the first time step! */
  if (n==0) {
    lambda_max_memory=lambda_max;
    laim=lambda_max_memory;
  }
  // if (n==0){printf("> in LAI_KNORR: T_memory = %2.1f\n",T_memory);}
  // if (n==0){T_memory=T_phi+3*T_r;}   /* set the temperature memory to be high so that we start in growth phase */
  // Exponentially declining memory of temperature
  T      = exp(- 1 / tau_m)*T_memory + meantemp * (1 - exp(- 1 / tau_m));
  T_deviation=(T-T_phi)/T_r;
  /* fraction of plants above temperature threshold: using cumulative normal distribution function (derived from the intrinsic c-function erfc) */
  f_T    = 0.5*erfc(-T_deviation*sqrt(0.5));

  /* compute daylight hours (hours of sunlight per day) */

  daylength = ComputeDaylightHours(daylengthpars);
  td_deviation = (daylength-t_c)/t_r;
  /* fraction of plants above daylength threshold: using cumulative normal distribution function (derived from the intrinsic c-function erfc) */
  f_d    = 0.5*erfc(-td_deviation*sqrt(0.5));  

  /* fraction of plants above temperature and daylength thresholds */
  f      = f_T * f_d;
  /* compute current growth rate */
  r      = plgr * f + (1 - f)*k_L;

  /* compute water-limited maximum LAI */
  lambda_W    = (pasm * lambda) / (tau_W * MaxExponentialSmooth(ET, 1e-3, 2e-2));
  /* compute smoothed maximum LAI */
  lambda_tilde_max = MinQuadraticSmooth(lambda_max, lambda_W, 0.99);
  /* update LAI water/structural memory using an exponentially declining memory of water/structural limitation over the time period tau_s */
  laim = exp(- 1.0 / tau_s)*lambda_max_memory + lambda_tilde_max * (1.0 - exp(- 1.0 / tau_s));
  lambda_lim = MaxExponentialSmooth(plgr * laim * f / r, 1e-9, 5e-3);

  // lambda_next is the updated target LAI (units of m2/m2)
  lambda_next = lambda_lim - (lambda_lim - lambda)*exp(-r*deltat);

  // dlambdadt is the change in LAI over the period deltat (units of m2/m2)
  dlambdadt = (lambda_next - lambda);
  

  
  //Store outputs in struct
 K->OUT.lambda_next = lambda_next;
 K->OUT.T= T;
 K->OUT.laim = laim;
 K->OUT.dlambdadt = dlambdadt;//dLAI/dt units = change of LAI per timestep
 K->OUT.f_T=f_T;
 K->OUT.f_d= f_d;
 


 
 
  return 0;
}

File: PYTHON/projects/CARDAMOM_MODELS/DALEC/DALEC_ALL/KNORR_ALLOCATION.py
import jax.numpy as jnp
from jax.scipy.special import erfc

# =====================================================================
# HELPER FUNCTIONS (From LAI_KNORR_funcs.c)
# =====================================================================
def MinQuadraticSmooth(x, y, eta):
    """Minimum function with quadratic transition (smoothed) between two values."""
    z = (x + y)**2.0 - 4.0 * eta * x * y
    z = jnp.maximum(z, 1e-18)
    mins = (x + y - jnp.sqrt(z)) / (2.0 * eta)
    return mins

def MaxExponentialSmooth(x, y, x0):
    """Maximum function with exponential transition (smoothed) between two values."""
    return jnp.where(
        x >= (y - x0), 
        x + x0 * jnp.exp(-(x - y) / x0 - 1.0), 
        y
    )

def ComputeDaylightHours(latitude, DOY):
    """Day length function - hours of sunlight per day."""
    pi = jnp.pi
    
    # Declination angle
    dec = -23.4 * jnp.cos((360.0 * (DOY + 10.0) / 365.0) * pi / 180.0) * pi / 180.0
    mult = jnp.tan(latitude * pi / 180.0) * jnp.tan(dec)
    
    # Clip multiplier to [-1, 1] to avoid NaNs in arccos, replacing the C if/else bounds
    mult_clipped = jnp.clip(mult, -1.0, 1.0)
    dayl = 24.0 * jnp.arccos(-mult_clipped) / pi
    
    return dayl

# =====================================================================
# MAIN MODULE (From KNORR_ALLOCATION.c)
# =====================================================================
def KNORR_ALLOCATION(temp, deltat, latitude, DOY, lambda_lai, lambda_max, 
                     T_phi, T_r, plgr, k_L, pasm, transp, tau_W, t_c, t_r, 
                     T_memory, lambda_max_memory):
    """
    Differentiable JAX translation of the Knorr Leaf Phenology Model.
    Calculates the target LAI, growth rate, and updates phenology memories.
    """
    tau_m = 30.0 # Temperature memory averaging period
    tau_s = 30.0 # Water/structural memory averaging period
    
    # -------------------------------------------------------------------------
    # 1. Temperature Threshold (f_T)
    # -------------------------------------------------------------------------
    # Exponentially declining memory of temperature
    T = jnp.exp(-deltat / tau_m) * T_memory + temp * (1.0 - jnp.exp(-deltat / tau_m))
    T_deviation = (T - T_phi) / T_r
    
    # Fraction of plants above temp threshold using cumulative normal distribution
    f_T = 0.5 * erfc(-T_deviation * jnp.sqrt(0.5))
    
    # -------------------------------------------------------------------------
    # 2. Daylength Threshold (f_d)
    # -------------------------------------------------------------------------
    daylength = ComputeDaylightHours(latitude, DOY)
    td_deviation = (daylength - t_c) / t_r
    
    # Fraction of plants above daylength threshold
    f_d = 0.5 * erfc(-td_deviation * jnp.sqrt(0.5))
    
    # -------------------------------------------------------------------------
    # 3. Growth Rates and Smoothing
    # -------------------------------------------------------------------------
    # Fraction of plants above both thresholds
    f = f_T * f_d
    
    # Current growth rate
    r = plgr * f + (1.0 - f) * k_L
    
    # Compute water-limited max LAI (added 1e-9 to prevent divide by zero)
    transp_smooth = MaxExponentialSmooth(transp, 1e-3, 2e-2)
    lambda_W = (pasm * lambda_lai) / (tau_W * transp_smooth + 1e-9)
    
    # Smoothed max LAI
    lambda_tilde_max = MinQuadraticSmooth(lambda_max, lambda_W, 0.99)
    
    # Update LAI structural memory
    laim = jnp.exp(-deltat / tau_s) * lambda_max_memory + lambda_tilde_max * (1.0 - jnp.exp(-deltat / tau_s))
    
    # Limiting target LAI
    lambda_lim = MaxExponentialSmooth(plgr * laim * f / (r + 1e-9), 1e-9, 5e-3)
    
    # -------------------------------------------------------------------------
    # 4. Final Updates
    # -------------------------------------------------------------------------
    # Updated target LAI (units of m2/m2)
    lambda_next = lambda_lim - (lambda_lim - lambda_lai) * jnp.exp(-r * deltat)
    
    # Change in LAI over deltat
    dlambdadt = lambda_next - lambda_lai

    return (lambda_next, T, laim, dlambdadt, f_T, f_d, lambda_tilde_max, lambda_W)
